clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,178 +23,419 @@ package common
+			if status := checker.getStatus(); status {
+				numberRunning++
+			}
+		}
+		w.lock.Unlock()
+		if !isChecking && isUpToDate && desiredCount == numberRunning {
+			break
+		}
+	}
+
+	if isChecking || !isUpToDate || desiredCount != numberRunning {
+		return fmt.Errorf("timed out while waiting for controlled pods")
+	}
+	glog.Infof("%s: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, obj runtime.Object) {
+	objName, err := getNameFromRuntimeObject(obj)
+	if err != nil {
+		glog.Errorf("%s: reading object name error: %v", waitForControlledPodsRunningName, err)
+		return
+	}
+	objNamespace, err := getNamespaceFromRuntimeObject(obj)
+	if err != nil {
+		glog.Errorf("%s: reading object namespace error: %v", waitForControlledPodsRunningName, err)
+		return
+	}
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if !w.isRunning {
+		return
+	}
+	key := createMetaNamespaceKey(objNamespace, objName)
+	if checker, exists := w.checkerMap[key]; exists {
+		checker.terminate()
+	}
+	checker, err := w.waitForRuntimeObject(c, obj)
+	if err != nil {
+		glog.Errorf("%s: waiting for %v (%v) error: %v", waitForControlledPodsRunningName, err)
+		return
+	}
+	time.AfterFunc(w.operationTimeout, func() {
+		checker.terminate()
+	})
+	w.checkerMap[key] = checker
+}
+
+func (w *waitForControlledPodsRunningMeasurement) deleteObject(obj runtime.Object) {
+	objName, err := getNameFromRuntimeObject(obj)
+	if err != nil {
+		glog.Errorf("%s: reading object name error: %v", waitForControlledPodsRunningName, err)
+		return
+	}
+	objNamespace, err := getNamespaceFromRuntimeObject(obj)
+	if err != nil {
+		glog.Errorf("%s: reading object namespace error: %v", waitForControlledPodsRunningName, err)
+		return
+	}
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if !w.isRunning {
+		return
+	}
+	glog.Infof("%s: %v (%v) has been deleted", waitForControlledPodsRunningName, objName, objNamespace)
+	key := createMetaNamespaceKey(objNamespace, objName)
+	if checker, exists := w.checkerMap[key]; exists {
+		checker.terminate()
+		delete(w.checkerMap, key)
 	}
-	return summaries, fmt.Errorf("error while waiting for %ss to be running in namespace '%v' with labels '%v' and fields '%v' - only %d found running with all pods: %v",
-		kind, namespace, labelSelector, fieldSelector, runningRuntimeObjects, errList.String())
 }
 
-func waitForRuntimeObject(c clientset.Interface, obj runtime.Object, timeout time.Duration) error {
+func (w *waitForControlledPodsRunningMeasurement) updateOpResourceVersion(resourceVersion string) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	version, err := strconv.ParseUint(resourceVersion, 10, 64)
```

Comment:
Please opaque this into function.

Also this is the function you want to copy:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd/api_object_versioner.go#L73

BTW - you should also copy the last one for comparing from that file too.

Reply:
I copied `ObjectResourceVersion`. I think comparing function will not be needed.